File: python/mortgageRate/model/metamorphic_relation.py
from abc import ABC, abstractmethod
from typing import List

from python.mortgageRate.model.metamorphic_group import MetamorphicGroup
from python.mortgageRate.utils.load_mrs_utils import load_all_metamorphic_relations
from python.mortgageRate.model.test_case import TestCase


class MetamorphicRelation(ABC):
    """
    蜕变关系的抽象基类

    定义了蜕变关系应实现的接口，包括后续测试用例生成和关系验证

    所有具体的蜕变关系类都应该继承这个基类并实现其抽象方法
    """

    @abstractmethod
    def get_id(self) -> str:
        """
        获取蜕变关系的ID

        Returns:
            str: 蜕变关系ID (例如 "mr_1")
        """
        pass

    @abstractmethod
    def get_description(self) -> str:
        """
        获取蜕变关系的描述

        Returns:
            str: 蜕变关系的人类可读描述
        """
        pass

    @abstractmethod
    def generate_followup_tests(self, source_test: TestCase) -> List[TestCase]:
        """
        基于源测试用例生成后续测试用例

        Args:
            source_test (test_case): 源测试用例

        Returns:
            List[TestCase]: 生成的后续测试用例列表
        """
        pass

    @abstractmethod
    def verify_relation(self, source_test: TestCase, followup_test: TestCase,
                        source_result: float, followup_result: float,
                        source_execution: str = "", followup_execution: str = "") -> bool:
        """
        验证源测试和后续测试的结果是否满足蜕变关系

        Args:
            source_test (test_case): 源测试用例
            followup_test (test_case): 后续测试用例
            source_result (float): 源测试结果
            followup_result (float): 后续测试结果
            source_execution (str, optional): 源测试执行的错误信息，如果有的话
            followup_execution (str, optional): 后续测试执行的错误信息，如果有的话

        Returns:
            bool: 如果满足蜕变关系返回True，否则返回False
        """
        pass

    @abstractmethod
    def is_applicable_to(self, test_case: TestCase) -> bool:
        """
        检查蜕变关系是否适用于给定的测试用例

        Args:
            test_case (test_case): 要检查的测试用例

        Returns:
            bool: 如果蜕变关系适用于给定的测试用例返回True，否则返回False
        """
        pass

    def create_groups(self, source_test: TestCase) -> List[MetamorphicGroup]:
        """
        基于源测试用例创建蜕变组

        Args:
            source_test (test_case): 源测试用例

        Returns:
            List[metamorphic_group]: 创建的蜕变组列表
        """
        groups = []

        # 检查蜕变关系是否适用于源测试用例
        if not self.is_applicable_to(source_test):
            return groups

        # 生成后续测试用例
        followup_tests = self.generate_followup_tests(source_test)

        # 为每个后续测试用例创建一个蜕变组
        for followup_test in followup_tests:
            group = MetamorphicGroup(
                self.get_id(),
                self.get_description(),
                source_test,
                followup_test
            )
            groups.append(group)

        return groups


class MetamorphicRelationFactory:
    """
    蜕变关系工厂类

    提供用于创建和管理蜕变关系的静态方法
    """

    @staticmethod
    def create_all_relations() -> List[MetamorphicRelation]:
        """
        创建所有支持的蜕变关系

        Returns:
            List[MetamorphicRelation]: 所有蜕变关系的列表
        """
        # 创建并返回所有蜕变关系的实例
        return load_all_metamorphic_relations()

    @staticmethod
    def generate_metamorphic_groups(test_case: TestCase) -> List[MetamorphicGroup]:
        """
        为测试用例生成所有适用的蜕变组

        Args:
            test_case (test_case): 源测试用例

        Returns:
            List[metamorphic_group]: 生成的蜕变组列表
        """
        groups = []

        for relation in MetamorphicRelationFactory.create_all_relations():
            if relation.is_applicable_to(test_case):
                groups.extend(relation.create_groups(test_case))

        return groups

File: java/src/main/java/paper/pss/exp/jfreeChart_project/model/MetamorphicRelation.java
package paper.pss.exp.jfreeChart_project.model;

import java.awt.Shape;
import java.util.List;

/**
 * MetamorphicRelation.java
 *
 * 这个接口定义了用于测试createLineRegion函数的蜕变关系。
 * 每个蜕变关系描述了如何在输入发生变化时输出应该如何变化，并提供了
 * 生成后续测试用例和验证关系属性的方法。
 *
 * 框架包括六个蜕变关系：
 * - MR1: 宽度比例关系 - 检查宽度变化时区域面积的变化比例
 * - MR2: 线段长度比例关系 - 检查线段长度变化时区域面积的变化比例
 * - MR3: 线段对称性关系 - 验证交换线段端点后形状的等价性
 * - MR4: 平移不变性关系 - 验证平移操作后形状特性的保持
 * - MR5: 正交特性关系 - 验证水平和垂直线段的特殊性质
 * - MR6: 旋转不变性关系 - 验证旋转操作后形状特性的保持
 */
public interface MetamorphicRelation {

    /**
     * 获取关系标识符
     *
     * @return 关系ID (例如, "MR1")
     */
    String getId();

    /**
     * 获取关系的可读描述
     *
     * @return 描述文本
     */
    String getDescription();

    /**
     * 根据源测试用例生成后续测试用例
     *
     * @param sourceTest 原始测试用例
     * @return 生成的后续测试用例列表
     */
    List<TestCase> generateFollowupTests(TestCase sourceTest);

    /**
     * 验证测试结果是否满足蜕变关系
     *
     * @param sourceTest        源测试用例
     * @param followupTest      后续测试用例
     * @param sourceResult      源测试结果
     * @param followupResult    后续测试结果
     * @param sourceExecution   源执行中的错误消息/类型 (若无错误则为空)
     * @param followupExecution 后续执行中的错误消息/类型 (若无错误则为空)
     * @return 如果关系满足则返回true
     */
    boolean verifyRelation(TestCase sourceTest, TestCase followupTest,
            Shape sourceResult, Shape followupResult,
            String sourceExecution, String followupExecution);

    /**
     * 判断此关系是否适用于给定的测试用例
     *
     * @param testCase 要检查的测试用例
     * @return 如果关系适用则返回true
     */
    boolean isApplicableTo(TestCase testCase);

    /**
     * 为给定的源测试用例创建蜕变组
     *
     * @param sourceTest 源测试用例
     * @return 蜕变组列表 (每个后续测试一个)
     */
    List<MetamorphicGroup> createGroups(TestCase sourceTest);


}